cmd/plugins/memory-qos/main.go

```diff
@@ -0,0 +1,304 @@
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/containerd/nri/pkg/api"
+	"github.com/containerd/nri/pkg/stub"
+)
+
+type plugin struct {
+	stub   stub.Stub
+	mask   stub.EventMask
+	config *pluginConfig
+}
+
+type pluginConfig struct {
+	// UnifiedAnnotations lists keys whose values are written
+	// directly from annotations to the OCI Linux unified
+	// object. Example:
+	//     UnifiedAnnotations: ["memory.high", "memory.swap.max"]
+	// allows using pod annotation
+	//     memory.swap.max.memory-qos.nri.io: max
+	// that will add unified["memory.swap.max"] = "max"
+	UnifiedAnnotations []string
+
+	// Classes define how memory of all workloads in each QoS
+	// class should be managed.
+	Classes []QoSClass
+}
+
+type QoSClass struct {
+	// Name of the QoS class, matches to annotations in
+	// pods. Examples:
+	// Set the default class for containers in the pod:
+	//     class.memory-qos.nri.io: "swap"
+	// Override the default class of CONTAINERNAME:
+	//     class.memory-qos.nri.io/CONTAINERNAME: "noswap"
+	Name string
+
+	// SwapLimitRatio sets memory.high based on memory limit.
+	// 1.0 means no throttling before getting OOM-killed.
+	// 0.75 throttle (reclaim pages) when usage reaches 75 % of memory limit.
+	SwapLimitRatio float32
+}
+
+const (
+	annotationSuffix = ".memory-qos.nri.io"
+)
+
+var (
+	log *logrus.Logger
+)
+
+// Configure handles connecting to container runtime's NRI server.
+func (p *plugin) Configure(ctx context.Context, config, runtime, version string) (stub.EventMask, error) {
+	log.Infof("Connected to %s %s...", runtime, version)
+	if config != "" {
+		log.Debugf("loading configuration from NRI server")
+		if err := p.setConfig([]byte(config)); err != nil {
+			return 0, err
+		}
+		return 0, nil
+	}
+	return 0, nil
+}
+
+// onClose handles losing connection to container runtime.
+func (p *plugin) onClose() {
+	log.Infof("Connection to the runtime lost, exiting...")
+	os.Exit(0)
+}
+
+// setConfig applies new plugin configuration.
+func (p *plugin) setConfig(config []byte) error {
+	log.Tracef("setConfig: parsing\n---8<---\n%s\n--->8---", config)
+	cfg := pluginConfig{}
+	err := yaml.Unmarshal(config, &cfg)
+	if err != nil {
+		errWithContext := fmt.Errorf("setConfig: cannot parse configuration: %w", err)
+		log.Debugf("%s", errWithContext)
+		return errWithContext
+	}
+	p.config = &cfg
+	log.Tracef("new configuration has %d classes:", len(p.config.Classes))
+	for _, cls := range p.config.Classes {
+		log.Tracef("- %s", cls.Name)
+	}
+	return nil
+}
+
+// pprintCtr() returns human readable container name that is
+// unique to the node.
+func pprintCtr(pod *api.PodSandbox, ctr *api.Container) string {
+	return fmt.Sprintf("%s/%s:%s", pod.GetNamespace(), pod.GetName(), ctr.GetName())
+}
+
+// applyQosClass applies QoS class to a container, updates unified values.
```

Comment:
```suggestion
// applyQosClass applies memory QoS class to a container, updates unified values.
```

Reply:
Fixed this and all other places, too... now the documentation always talks about "memory QoS classes" or "classes" but never "QoS classes" to avoid confusion with guaranteed/burstable/besteffort.